unicorn/include/cocaine/idl/unicorn.hpp

```diff
@@ -0,0 +1,267 @@
+* the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*/
+
+#ifndef COCAINE_UNICORN_SERVICE_INTERFACE_HPP
+#define COCAINE_UNICORN_SERVICE_INTERFACE_HPP
+
+#include "cocaine/unicorn/path.hpp"
+#include "cocaine/unicorn/value.hpp"
+
+#include <cocaine/rpc/protocol.hpp>
+#include <boost/mpl/list.hpp>
+#include <vector>
+
+namespace cocaine { namespace io {
+
+struct unicorn_tag;
+struct unicorn_final_tag;
+struct unicorn_locked_tag;
+
+using namespace cocaine::unicorn;
+
+struct unicorn {
+    struct create {
+        typedef unicorn_tag tag;
+
+        static const char* alias() {
+            return "create";
+        }
+
+        /**
+        * Create a node if it does not exist.
+        *
+        * path_t - path to change.
+        * value_t - value to write in path
+        **/
+        typedef boost::mpl::list<
+            path_t,
+            value_t
+        > argument_type;
+
+        /**
+        * true if node was created. Error on any kind of error
+        */
+        typedef option_of<
+            bool
+        >::tag upstream_type;
+
+        typedef unicorn_final_tag dispatch_type;
+    };
+
+    struct put {
+        typedef unicorn_tag tag;
+
+        static const char* alias() {
+            return "put";
+        }
+
+        /**
+        *  put command implements cas behaviour. It accepts:
+        *
+        * path_t - path to change.
+        * value_t - value to write in path
+        * version_t - version to compare with. If version in zk do not match - error will be returned.
+        *   -1 indicates that version check is not performed (forced put).
+        **/
+        typedef boost::mpl::list<
+            path_t,
+            value_t,
+            version_t
+        > argument_type;
+
+        /**
+        * Return value is current value in ZK.
+        */
+        typedef option_of<
+            bool,
+            versioned_value_t
+        >::tag upstream_type;
+
+        typedef unicorn_final_tag dispatch_type;
+    };
+
+    struct subscribe {
+        typedef unicorn_tag tag;
+
+        static const char* alias() {
+            return "subscribe";
+        }
+
+        /**
+        * subscribe for updates on path. Will send last update which version is greater than specified.
+        */
+        typedef boost::mpl::list<
+            path_t
+        > argument_type;
+
+        /**
+        * current version in ZK
+        */
+        typedef stream_of<
+            versioned_value_t
+        >::tag upstream_type;
+
+        typedef unicorn_final_tag dispatch_type;
+    };
+
+    struct del {
+        typedef unicorn_tag tag;
+
+        static const char* alias() {
+            return "del";
```

Comment:
Strong concordance between the struct name and its alias isn't required. If you want to name your method as `delete` do so.